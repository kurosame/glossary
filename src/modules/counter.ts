import { Reducer } from 'redux'
import {
  Action,
  ActionFunctionAny,
  createActions,
  handleActions
} from 'redux-actions'

// Action types
export const ADD_COUNT = 'ADD_COUNT'
export const ADD_SAGA_COUNT = 'ADD_SAGA_COUNT'
export const GET_SAGA_COUNT = 'GET_SAGA_COUNT'

export interface ICounterActions {
  addCount: ActionFunctionAny<Action<{}>>
  addSagaCount: ActionFunctionAny<Action<{ sagaCount: number }>>
  getSagaCount: ActionFunctionAny<Action<{}>>
}

export interface ICounterState {
  count: number
  sagaCount: number
}

const initialState: ICounterState = {
  count: 0,
  sagaCount: 0
}

export const { addCount, addSagaCount, getSagaCount } = createActions(
  ADD_COUNT,
  ADD_SAGA_COUNT,
  GET_SAGA_COUNT
)

export const counter: Reducer = handleActions(
  {
    [ADD_COUNT]: (state: ICounterState): ICounterState => {
      return { ...state, count: state.count + 1 }
    },
    [ADD_SAGA_COUNT]: (
      state: ICounterState,
      action: Action<ICounterState>
    ): ICounterState => {
      return {
        ...state,
        sagaCount:
          state.sagaCount + (action.payload ? action.payload.sagaCount : 0)
      }
    }
  },
  initialState
)
