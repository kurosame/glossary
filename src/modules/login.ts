import { Reducer } from 'redux'
import {
  Action,
  ActionFunctionAny,
  createActions,
  handleActions
} from 'redux-actions'

// Action types
export const SET_IS_LOGIN = 'SET_IS_LOGIN'

export interface LoginActions {
  setIsLogin: ActionFunctionAny<Action<{ isLogin: boolean }>>
}

export interface LoginState {
  isLogin: boolean
}

const initialState: LoginState = { isLogin: false }

export const { setIsLogin } = createActions<{ isLogin: boolean }>(SET_IS_LOGIN)

export const login: Reducer<
  LoginState,
  Action<{ isLogin: boolean }>
> = handleActions(
  {
    [SET_IS_LOGIN]: (
      _state: LoginState,
      action: Action<{ isLogin: boolean }>
    ): LoginState => {
      return { isLogin: action.payload.isLogin }
    }
  },
  initialState
)
