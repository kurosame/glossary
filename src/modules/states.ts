import type { LoginState } from '@/modules/login'
import type { WordState } from '@/modules/word'

export interface States {
  login: LoginState
  words: WordState[]
}
