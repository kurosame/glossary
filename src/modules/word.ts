import { Reducer } from 'redux'
import {
  Action,
  ActionFunctionAny,
  createActions,
  handleActions
} from 'redux-actions'

// Action types
export const GET_WORDS = 'GET_WORDS'
export const SET_WORDS = 'SET_WORDS'

export interface IWordActions {
  getWords: ActionFunctionAny<Action<{}>>
  setWords: ActionFunctionAny<Action<{ words: IWordState[] }>>
}

export interface IWordState {
  id: string
  category: string
  titles: string[]
  description: string
}

const initialState: IWordState[] = []

export const { getWords, setWords } = createActions(GET_WORDS, SET_WORDS)

export const words: Reducer<IWordState[], Action<IWordState[]>> = handleActions(
  {
    [SET_WORDS]: (
      state: IWordState[],
      action: Action<IWordState[]>
    ): IWordState[] => {
      return !!action.payload.length ? action.payload : state
    }
  },
  initialState
)
