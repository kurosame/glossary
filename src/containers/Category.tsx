import { AppBar, Tab, Tabs } from '@material-ui/core'
import React from 'react'
import { useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { States } from '@/modules/states'

interface Props {
  location: { pathname: string }
}

const tabItems: Array<{ label: string; to: string }> = [
  { label: 'Summary', to: '/' },
  { label: 'Apache', to: '/apache' },
  { label: 'Architecture', to: '/architecture' },
  { label: 'AWS', to: '/aws' },
  { label: 'Browser', to: '/browser' },
  { label: 'CI', to: '/ci' },
  { label: 'Compiler', to: '/compiler' },
  { label: 'CSS', to: '/css' },
  { label: 'Design', to: '/design' },
  { label: 'Docker', to: '/docker' },
  { label: 'Firebase', to: '/firebase' },
  { label: 'Go', to: '/go' },
  { label: 'Hardware', to: '/hardware' },
  { label: 'HTML', to: '/html' },
  { label: 'JS', to: '/js' },
  { label: 'Language', to: '/language' },
  { label: 'ML', to: '/ml' },
  { label: 'Mobile', to: '/mobile' },
  { label: 'Module Bundler', to: '/module-bundler' },
  { label: 'Network', to: '/network' },
  { label: 'Package Manager', to: '/package-manager' },
  { label: 'Programming', to: '/programming' },
  { label: 'Python', to: '/python' },
  { label: 'React', to: '/react' },
  { label: 'Scala', to: '/scala' },
  { label: 'Server', to: '/server' },
  { label: 'Swift', to: '/swift' },
  { label: 'Vue.js', to: '/vue' },
  { label: 'Words', to: '/words' },
  { label: 'Workflow Engine', to: '/workflow-engine' }
]

const Category: React.FC<Props> = p => {
  const isLogin = useSelector<States, boolean>(s => s.login.isLogin)

  if (!isLogin) return null
  return (
    <AppBar position="static">
      <Tabs
        value={tabItems.findIndex(o => o.to === p.location.pathname)}
        variant="scrollable"
        scrollButtons="off"
      >
        {tabItems.map(t => {
          const StaticLink = React.forwardRef((pp, ref: React.Ref<Link>) => (
            <Link ref={ref} to={t.to} {...pp} />
          ))
          StaticLink.displayName = 'StaticLink'
          return <Tab key={t.label} label={t.label} component={StaticLink} />
        })}
      </Tabs>
    </AppBar>
  )
}

export default Category
