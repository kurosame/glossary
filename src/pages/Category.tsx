import { AppBar, Tab, Tabs } from '@material-ui/core'
import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { States } from '@/modules/states'
import { LoginState } from '@/modules/login'

interface Props {
  state: { login: LoginState }
  location: { pathname: string }
}

const tabItems: Array<{ label: string; to: string }> = [
  { label: 'Summary', to: '/' },
  { label: 'Apache', to: '/apache' },
  { label: 'Architecture', to: '/architecture' },
  { label: 'AWS', to: '/aws' },
  { label: 'Browser', to: '/browser' },
  { label: 'CI', to: '/ci' },
  { label: 'Compiler', to: '/compiler' },
  { label: 'CSS', to: '/css' },
  { label: 'Design', to: '/design' },
  { label: 'Docker', to: '/docker' },
  { label: 'Firebase', to: '/firebase' },
  { label: 'Go', to: '/go' },
  { label: 'Hardware', to: '/hardware' },
  { label: 'HTML', to: '/html' },
  { label: 'JS', to: '/js' },
  { label: 'Language', to: '/language' },
  { label: 'ML', to: '/ml' },
  { label: 'Mobile', to: '/mobile' },
  { label: 'Module Bundler', to: '/module-bundler' },
  { label: 'Network', to: '/network' },
  { label: 'Package Manager', to: '/package-manager' },
  { label: 'Programming', to: '/programming' },
  { label: 'Python', to: '/python' },
  { label: 'React', to: '/react' },
  { label: 'Scala', to: '/scala' },
  { label: 'Server', to: '/server' },
  { label: 'Swift', to: '/swift' },
  { label: 'Vue.js', to: '/vue' },
  { label: 'Words', to: '/words' },
  { label: 'Workflow Engine', to: '/workflow-engine' }
]

const Category = (props: Props): JSX.Element | null =>
  props.state.login.isLogin ? (
    <AppBar position="static">
      <Tabs
        value={tabItems.findIndex(o => o.to === props.location.pathname)}
        scrollable={true}
        scrollButtons="off"
        data-test="category-tabs"
      >
        {tabItems.map(t => (
          <Tab
            key={t.label}
            label={t.label}
            component={(p: {}): JSX.Element => <Link to={t.to} {...p} />}
          />
        ))}
      </Tabs>
    </AppBar>
  ) : null

export default connect((states: States) => ({
  state: { login: states.login }
}))(Category)
