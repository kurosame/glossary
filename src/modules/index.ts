import { ICounterActions, ICounterState } from '@/modules/counter'
import { IWordActions, IWordState } from '@/modules/word'

export interface IActions {
  counter: ICounterActions
  word: IWordActions
}

export interface IStates {
  counter: ICounterState
  words: IWordState[]
}
