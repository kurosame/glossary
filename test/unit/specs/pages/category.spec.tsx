import { States } from '@/modules/states'
import Category from '@/containers/Category'
import { shallow } from 'enzyme'
import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import configureStore from 'redux-mock-store'

describe('Match the snapshot', () => {
  test('The isLogin is true', () => {
    const wrapper = shallow(
      <Router initialEntries={['/']}>
        <Category
          location={{ pathname: '/react' }}
          store={configureStore<Pick<States, 'login'>>()({
            login: { isLogin: true }
          })}
        />
      </Router>
    )

    expect(wrapper.html()).not.toEqual('')
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('The isLogin is false', () => {
    const wrapper = shallow(
      <Router initialEntries={['/']}>
        <Category
          location={{ pathname: '/react' }}
          store={configureStore<Pick<States, 'login'>>()({
            login: { isLogin: false }
          })}
        />
      </Router>
    )

    expect(wrapper.html()).toEqual('')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
