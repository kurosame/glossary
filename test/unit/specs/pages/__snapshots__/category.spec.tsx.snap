// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match the snapshot The isLogin is false 1`] = `""`;

exports[`Match the snapshot The isLogin is true 1`] = `"<header class=\\"MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4\\"><div class=\\"MuiTabs-root\\" data-test=\\"category-tabs\\"><div style=\\"width:99px;height:99px;position:absolute;top:-9999px;overflow:scroll\\" class=\\"MuiTabs-scrollable\\"></div><div class=\\"MuiTabs-scroller MuiTabs-scrollable\\" style=\\"overflow:hidden\\"><div class=\\"MuiTabs-flexContainer\\" role=\\"tablist\\"><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/\\"><span class=\\"MuiTab-wrapper\\">Summary</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/apache\\"><span class=\\"MuiTab-wrapper\\">Apache</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/architecture\\"><span class=\\"MuiTab-wrapper\\">Architecture</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/aws\\"><span class=\\"MuiTab-wrapper\\">AWS</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/browser\\"><span class=\\"MuiTab-wrapper\\">Browser</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/ci\\"><span class=\\"MuiTab-wrapper\\">CI</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/compiler\\"><span class=\\"MuiTab-wrapper\\">Compiler</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/css\\"><span class=\\"MuiTab-wrapper\\">CSS</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/design\\"><span class=\\"MuiTab-wrapper\\">Design</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/docker\\"><span class=\\"MuiTab-wrapper\\">Docker</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/firebase\\"><span class=\\"MuiTab-wrapper\\">Firebase</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/go\\"><span class=\\"MuiTab-wrapper\\">Go</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/hardware\\"><span class=\\"MuiTab-wrapper\\">Hardware</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/html\\"><span class=\\"MuiTab-wrapper\\">HTML</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/js\\"><span class=\\"MuiTab-wrapper\\">JS</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/language\\"><span class=\\"MuiTab-wrapper\\">Language</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/ml\\"><span class=\\"MuiTab-wrapper\\">ML</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/mobile\\"><span class=\\"MuiTab-wrapper\\">Mobile</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/module-bundler\\"><span class=\\"MuiTab-wrapper\\">Module Bundler</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/network\\"><span class=\\"MuiTab-wrapper\\">Network</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/package-manager\\"><span class=\\"MuiTab-wrapper\\">Package Manager</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/programming\\"><span class=\\"MuiTab-wrapper\\">Programming</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/python\\"><span class=\\"MuiTab-wrapper\\">Python</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit Mui-selected\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"true\\" href=\\"/react\\"><span class=\\"MuiTab-wrapper\\">React</span><span class=\\"PrivateTabIndicator-root-62 PrivateTabIndicator-colorSecondary-64 MuiTabs-indicator\\"></span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/scala\\"><span class=\\"MuiTab-wrapper\\">Scala</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/server\\"><span class=\\"MuiTab-wrapper\\">Server</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/swift\\"><span class=\\"MuiTab-wrapper\\">Swift</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/vue\\"><span class=\\"MuiTab-wrapper\\">Vue.js</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/words\\"><span class=\\"MuiTab-wrapper\\">Words</span></a><a class=\\"MuiButtonBase-root MuiTab-root MuiTab-textColorInherit\\" tabindex=\\"0\\" role=\\"tab\\" aria-disabled=\\"false\\" aria-selected=\\"false\\" href=\\"/workflow-engine\\"><span class=\\"MuiTab-wrapper\\">Workflow Engine</span></a></div></div></div></header>"`;
